include/toolchain/xcc.h

```diff
@@ -38,4 +38,140 @@
 
 #define __builtin_unreachable() __ASSERT(0, "Unreachable code")
 
-#endif
+#define __builtin_add_overflow(a, b, output) ({ \
+	_Bool __ret__ = 0; \
+	if (__builtin_types_compatible_p(__typeof__(a), int)) { \
+		__ret__ = __builtin_sadd_overflow(a, b, output); \
+	} else if (__builtin_types_compatible_p(__typeof__(a), \
+						long int)) { \
+		__ret__ = __builtin_saddl_overflow(a, b, output); \
+	} else if (__builtin_types_compatible_p(__typeof__(a), \
+						unsigned int)) { \
+		__ret__ = __builtin_uadd_overflow(a, b, output); \
+	} else if (__builtin_types_compatible_p(__typeof__(a), \
+						unsigned long int)) { \
+		__ret__ = __builtin_uaddl_overflow(a, b, output); \
+	} else { \
+		/* assume overflow for all other types */ \
+		__ret__ = 1; \
+	} \
+	__ret__;\
+})
+
+#define __builtin_sadd_overflow(a, b, output) ({ \
```

Comment:
Oh, xcc doesn't even have these?  You can maybe simplify this by using a macro such as this:

```C
#define _xcc_bin_op_overflow(a, b, small_type, overflow_type, operator, output) \
  ({ \
          overflow_type result = a operator b; \
          *output = (small_type)result; \
          *output != (overflow_type)(small_type)result; \
   })
```

And then defining, for instance, `__builtin_sadd_overflow` to `_xcc_bin_op_overflow(a, b, i32_t, i64_t, +, output)`, and so on.  Better to use the explicit types as well (`i32_t` vs `int`), because sometimes, `int` and `long int` are the same, and you want the wider type to capture the overflow.  The same macro can be used for multiplication as well.

I don't know, however, how good XCC is here to know that you're only interested in knowing that the operation overflowed, so you might want to look at alternative methods of detecting overflow if the generated code is terrible when promoting 32 to 64 bit.